### PR TITLE
Add precondition format checks for AppId and API Key

### PR DESCRIPTION
**Pull request by ChaoqunCHEN**
Add strict format checks on AppId and API key.  Throws exception with solution if SDK detects incorrect format of AppId or API key.

Refer to https://github.com/firebase/firebase-android-sdk/blob/master/firebase-installations/REQUIRED_FIREBASE_OPTIONS_ANDROID.md for details.

### CODE REVIEW

File: firebase-installations/src/main/java/com/google/firebase/installations/Utils.java [hunk 1]
```diff
@@ -41,4 +45,12 @@ public boolean isAuthTokenExpired(PersistedInstallationEntry entry) {
   public long currentTimeInSecs() {
     return TimeUnit.MILLISECONDS.toSeconds(System.currentTimeMillis());
   }
+
+  static boolean isValidAppIdFormat(String AppId) {
```
Comment (andirayo) [inline firebase-installations/src/main/java/com/google/firebase/installations/Utils.java:49]: lowercase variable names:
`appId`

File: firebase-installations/src/main/java/com/google/firebase/installations/Utils.java [hunk 2]
```diff
@@ -41,4 +45,12 @@ public boolean isAuthTokenExpired(PersistedInstallationEntry entry) {
...
+    return AppId.contains(APP_ID_IDENTIFICATION_SUBSTRING);
+  }
+
+  static boolean isValidApiKeyFormat(String ApiKey) {
```
Comment (andirayo) [inline firebase-installations/src/main/java/com/google/firebase/installations/Utils.java:53]: lowercase variable names:
`apiKey`

I think we can also add a @NonNull here.

File: firebase-installations/src/main/java/com/google/firebase/installations/Utils.java (hunk 1 above)
Comment (andirayo) [reply firebase-installations/src/main/java/com/google/firebase/installations/Utils.java:49]: I think we can also add a @NonNull here.

File: firebase-installations/src/main/java/com/google/firebase/installations/Utils.java [hunk 3]
```diff
@@ -17,11 +17,15 @@
 import android.text.TextUtils;
 import com.google.firebase.installations.local.PersistedInstallationEntry;
 import java.util.concurrent.TimeUnit;
+import java.util.regex.Pattern;
```
Comment (andirayo) [inline firebase-installations/src/main/java/com/google/firebase/installations/Utils.java:20]: Please check with Android folks (e.g. Ashwin) if we can use that package here of if we should use re2j.

File: firebase-installations/src/test/java/com/google/firebase/installations/FirebaseInstallationsTest.java [hunk 4]
```diff
@@ -965,4 +966,41 @@ public void testDelete_networkError() throws Exception {
...
+    assertTrue(Utils.isValidApiKeyFormat("AIzaSyAnLA7NfeLquW1tJFpx_eQCxoX-oo6YyIs"));
+    // invalid ApiKey
+    assertFalse(
+        Utils.isValidApiKeyFormat("BIzaSyabcdefghijklmnopqrstuvwxyz1234567")); // wrong prefix
+    assertFalse(Utils.isValidApiKeyFormat("AIzaSyabcdefghijklmnopqrstuvwxyz")); // wrong length
+    assertFalse(Utils.isValidApiKeyFormat("AIzaSyabcdefghijklmno:qrstuvwxyzabcdefg")); // wrong char
+    assertFalse(Utils.isValidApiKeyFormat("AIzaSyabcdefghijklmno qrstuvwxyzabcdefg")); // wrong char
+    assertFalse(
+        Utils.isValidApiKeyFormat(
+            "AAAAdpB7anM:APA91bFFK03DIT8y3l5uymwbKcUDJdYqTRSP9Qcxg8SU5kKPalEpObdx0C0xv8gQttdWlLW4hLvvHA0JoDKA6Lrvbi-edUjFCPY_WJkuvHxFwGWXjnj4yI4sPQ27mXuSVIyAbgX4aTK0QYpIKq2j1NBi7ZU75gunQg")); // using FCM server key as API key.
```
Comment (andirayo) [inline firebase-installations/src/test/java/com/google/firebase/installations/FirebaseInstallationsTest.java:1004]: line length

File: firebase-installations/src/main/java/com/google/firebase/installations/FirebaseInstallations.java [hunk 5]
```diff
@@ -146,6 +146,14 @@ private void preConditionChecks() {
...
     Preconditions.checkNotEmpty(getProjectIdentifier());
     Preconditions.checkNotEmpty(getApiKey());
+    Preconditions.checkArgument(
+        Utils.isValidAppIdFormat(getApplicationId()),
+        "Invalid AppId format. Please refer to"
+            + " https://firebase.google.com/support/privacy/init-options for details.");
+    Preconditions.checkArgument(
+        Utils.isValidApiKeyFormat(getApiKey()),
+        "Invalid Api-key format. Please refer to"
+            + " https://firebase.google.com/support/privacy/init-options for details.");
```
Comment (andirayo) [inline firebase-installations/src/main/java/com/google/firebase/installations/FirebaseInstallations.java:156]: Please update similarly to IID change.

File: firebase-installations/src/main/java/com/google/firebase/installations/Utils.java (hunk 3 above)
Comment (ChaoqunCHEN) [reply firebase-installations/src/main/java/com/google/firebase/installations/Utils.java:20]: Greg said it's used in FCM SDK so it's fine.

File: firebase-installations/src/main/java/com/google/firebase/installations/Utils.java [hunk 6]
```diff
@@ -41,4 +48,12 @@ public boolean isAuthTokenExpired(PersistedInstallationEntry entry) {
   public long currentTimeInSecs() {
     return TimeUnit.MILLISECONDS.toSeconds(System.currentTimeMillis());
   }
+
+  static boolean isValidAppIdFormat(@Nullable String appId) {
```
Comment (andirayo) [inline firebase-installations/src/main/java/com/google/firebase/installations/Utils.java:52]: This should be @NonNull - not @Nullable.

File: firebase-installations/src/main/java/com/google/firebase/installations/Utils.java [hunk 7]
```diff
@@ -41,4 +48,12 @@ public boolean isAuthTokenExpired(PersistedInstallationEntry entry) {
...
+    return appId.contains(APP_ID_IDENTIFICATION_SUBSTRING);
+  }
+
+  static boolean isValidApiKeyFormat(@Nullable String apiKey) {
```
Comment (andirayo) [inline firebase-installations/src/main/java/com/google/firebase/installations/Utils.java:56]: This should be @NonNull - not @Nullable.

File: firebase-installations/src/main/java/com/google/firebase/installations/Utils.java (hunk 3 above)
Comment (andirayo) [reply firebase-installations/src/main/java/com/google/firebase/installations/Utils.java:23]: He might have said that because it's GMSCore and GMSCore has rules that apply to older code.
Maybe for other Firebase SDKs we have other requirements.
Please check with someone from Android team, if it is fine to use this package instead of `re2j`.

File: firebase-installations/src/main/java/com/google/firebase/installations/FirebaseInstallations.java [hunk 8]
```diff
@@ -143,18 +143,19 @@ public Thread newThread(Runnable r) {
...
-        getProjectIdentifier(),
-        "Please set your project ID. A valid Firebase project ID is required to communicate "
-            + "with Firebase server APIs: It identifies your project with Google.");
-    Preconditions.checkNotEmpty(
-        getApiKey(),
-        "Please set a valid API key. A Firebase API key is required to communicate with "
-            + "Firebase server APIs: It authenticates your project with Google.");
+    Preconditions.checkNotEmpty(getApplicationId());
+    Preconditions.checkNotEmpty(getProjectIdentifier());
+    Preconditions.checkNotEmpty(getApiKey());
```
Comment (andirayo) [inline firebase-installations/src/main/java/com/google/firebase/installations/FirebaseInstallations.java:148]: Why did you remove the messages?
This is important for developers to see!
Please re-add!
